### PR TITLE
RHOAIENG-8450: feat(odh-notebook-controller): add back notebook container envs var from central proxy configs

### LINKED ISSUE #291: add back missing default env vars from openshift proxy config, if present
/kind feature

**Why you need this feature:**
During the last round of updates related to CA trust, code was removed that looks for an [openshift central proxy config ](https://docs.openshift.com/container-platform/4.12/networking/enable-cluster-wide-proxy.html) and adds the centrally configured values for HTTP_PROXY, HTTPS_PROXY, NO_PROXY

This is what used to be in there before:
[186106](https://github.com/opendatahub-io/kubeflow/pull/43/files#diff-1df80e242be90c94fce3ffb051a0bfe706fe924dab62438c7aa0a186b8067153)

**Describe the solution you'd like:**
Take the old code, leave out trusted CA related stuff (no volume mounts, no setting CA trust related env vars as that happens elsewhere now.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by shalberd**
/fixes #291 

notebook containers have HTTP_PROXY, HTTPS_PROXY, NO_PROXY env vars injected if central cluster proxy config exists

## Description
During the last round of updates related to CA trust, code was removed that looks for an [openshift central proxy config ](https://docs.openshift.com/container-platform/4.12/networking/enable-cluster-wide-proxy.html)and adds the centrally configured values for HTTP_PROXY, HTTPS_PROXY, NO_PROXY

## How Has This Been Tested?
not tested yet, I'd take the built image from quay.io and test whether the three variables appear in my notebooks / workbench containers.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 1]
```diff
@@ -221,6 +224,29 @@ func InjectOAuthProxy(notebook *nbv1.Notebook, oauth OAuthConfig) error {
 	return nil
 }
 
+func (w *NotebookWebhook) ClusterWideProxyIsEnabled() bool {
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:227]:
```suggestion
func (w *NotebookWebhook) ClusterWideProxyIsEnabled() (map[string]string, bool)  {
```

I think that `proxyEnvVars` does not need to be a global variable, especially since every call to `ClusterWideProxyIsEnabled` creates its content anew. Can you return the proxy env vars from the function?

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 2]
```diff
@@ -221,6 +224,29 @@ func InjectOAuthProxy(notebook *nbv1.Notebook, oauth OAuthConfig) error {
...
+	proxyResourceList := &configv1.ProxyList{}
+	err := w.Client.List(context.TODO(), proxyResourceList)
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:229]: There's already some context available in

https://github.com/opendatahub-io/kubeflow/blob/7401baff76e07f87b75a9ae4bdbb69431144722f/components/odh-notebook-controller/controllers/notebook_webhook.go#L250-L251

why don't you pass it in as parameter and use that, instead of creating a TODO context?

**Review by jiridanek — APPROVED** [review]
This is adding back previously removed code. So it is IMO fair to ignore the possible "improvements" I suggested, or that others may come up with, and merge it as it is / as it was.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 1 above)
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:227]: Hi @jiridanek very good point in general. There are certainly improvements possible, I am unsure whether now (adding back previous functionality) or later. I or you could create a new ticket for tracking with your suggestion and link it to issue #291

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 2 above)
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:229]: also very good point, I am still learning about all the ins and outs / details of odh notebook controller and the possibilities in golang. For here, I was merely adding back @VaishnaviHire code from way back then.

**Comment by shalberd** [comment]
> This is adding back previously removed code. So it is IMO fair to ignore the possible "improvements" I suggested, or that others may come up with, and merge it as it is / as it was.

Agreed, for tracking, I created issue #344

**Comment by atheo89** [comment]
Hi @shalberd could you rebase your PR  when you get a chance? There have been many changes since it was opened :)

**Comment by shalberd** [comment]
Hi @atheo89 I don't get the message that my forked branch is out of date compared to base branch, so I cannot rebase.
I mean, it says here "this branch has no conflicts with the base branch"

https://stackoverflow.com/questions/69839124/update-branch-with-rebase-instead-of-merge

Is a rebase really necessary here?

**Comment by shalberd** [comment]
@atheo89 ok, I see where you're coming from ... the whole pull request image build process and so on.
I did a rebase with v1.7-branch and force-pushed again, so ok now. Tests and image build re-running as of now.

**Comment by jiridanek** [comment]
Here's a tracking jira for it, https://issues.redhat.com/browse/RHOAIENG-8450

/lgtm

**Comment by shalberd** [comment]
@harshad16 ready to merge? This PR here is exclusively about proxy env var and value injection

**Comment by harshad16** [comment]
> @harshad16 ready to merge? This PR here is exclusively about proxy env var and value injection

Hey @shalberd , i will get this prioritized for next week :+1:

**Comment by jiridanek** [comment]
/retitle RHOAIENG-8450: feat(odh-notebook-controller): add back notebook container envs var from central proxy configs

**Comment by shalberd** [comment]
shall I change this PR to v1.9 branch?

**Comment by jiridanek** [comment]
no, against the `main` branch, please

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 3]
```diff
@@ -227,6 +230,29 @@ func InjectOAuthProxy(notebook *nbv1.Notebook, oauth OAuthConfig) error {
...
+	if err != nil {
+		return false
+	}
+
+	for _, proxy := range proxyResourceList.Items {
+		if proxy.Name == "cluster" {
+			if proxy.Status.HTTPProxy != "" && proxy.Status.HTTPSProxy != "" &&
+				proxy.Status.NoProxy != "" {
+				// Update Proxy Env variables map
+				proxyEnvVars["HTTP_PROXY"] = proxy.Status.HTTPProxy
```
[the remark this answers is not in the document]
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:245]: > We need to postpone the env vars update to when the notebook is restarted by the user, to prevent unexpected restart.

If that is really necessary, I would not know how to achieve this. In my view, it is overkill, but I realize you sometimes have differing requirements with environments with long running notebook containers.

[the remark this answers is not in the document]
Comment (jiridanek) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:245]: I'm setting up a fake operator upgrade scenario with envtest and so far it looks like that changing the envvars here does not modify the existing statefulsets. That seems to be because webhooks run only on create, update (and delete) of the notebook spec, and there's nothing else changing the spec, so the webhook is not rerun when operator restarts.

We have that one Openshift 4.13 cluster that does recreate pods where other clusters would not, so we need to figure out that case, but I'm hopeful this can be done without restarts somehow.

[the remark this answers is not in the document]
Comment (jiridanek) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:245]: Here are two proposed solutions to the notebook restart problem due to updates to the values that odh-notebook-controller is setting in the webhook

* https://github.com/opendatahub-io/kubeflow/pull/436
* https://github.com/opendatahub-io/kubeflow/pull/437

[the remark this answers is not in the document]
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:245]: @jiridanek is this still an issue? the restart issue. I lost track a little over time.

[the remark this answers is not in the document]
Comment (jiridanek) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:245]: This should be fixed by 

* https://github.com/opendatahub-io/kubeflow/pull/436

> Notebook CRs have a new annotation `notebooks.opendatahub.io/update-pending` that is added when there is a mismatch between the currently running Workbench and the desired configuration, and applying the desired configuration requires restarting the Workbench. Only one configuration change is displayed, even though there may be more than one pending updates. When there are no pending updates, for example when the Workbench is restarted and pending updates get applied on that occasion, the annotation is again removed by the odh-notebook-controller webhook.
>
> Example:

```
notebooks.opendatahub.io/update-pending: '{v1.PodSpec}.Containers[1].Image: registry.redhat.io/openshift4/ose-oauth-proxy@sha256:4f8d66597feeb32bb18699326029f9a71a5aca4a57679d636b876377c2e95695 != registry.redhat.io/openshift4/ose-oauth-proxy@sha256:4bef31eb993feb6f1096b51b4876c65a6fb1f4401fee97fa4f4542b6b7c9bc46'
```

> https://issues.redhat.com/browse/RHOAIENG-15335

There isn't a good automated test for this scenario yet, so this upgrade is one of the things I plan to try by hand, to be sure (I proposed https://github.com/opendatahub-io/kubeflow/pull/430 as automated test for upgrades, but the size of the accompanying YAMLs there scared everyone, so this is not going to be a way to do it).

**Comment by shalberd** [comment]
@jiridanek @atheo89 this is a little related to the notebook-images R studio env var PR ...
I rebased this PR to be up to date with the latest main branch changes in kubeflow itself.
Let me know if this is still ok from your perspective.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:248]: @jiridanek @jstourac I think you recently added changes to the webhook that take care of the issue.
https://github.com/opendatahub-io/kubeflow/commit/1a1894b8642f7b31492bcef07c6e626a5032fe76
So closing this as resolved.

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 4]
```diff
@@ -290,9 +324,9 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
 		mutatedNotebook.ObjectMeta.Annotations[updatePendingAnnotation] = needsRestart.Reason
 	} else {
 		delete(mutatedNotebook.ObjectMeta.Annotations, updatePendingAnnotation)
-	}
-
-	// Create the mutated notebook object
+        }
+	
+        // Create the mutated notebook object
```
Comment (shalberd) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:329]: @jiridanek no idea why the diff tools thinks there is a diff here ... all I did was rebase with main and resolve some merge conflicts

**Comment by jiridanek** [comment]
Woot! Coverage! 🎉

**Comment by shalberd** [comment]
> Woot! Coverage!

very nice, I will have to make myself more familiar with the findings.

Sven

**Comment by jiridanek** [comment]
> @jiridanek If you have an explanatory Web page somewhere on this topic of coverage, I'd appreciate it in the interest of learning more about it.

Sure, try these

1. https://about.codecov.io/resource/what-is-code-coverage/
1. https://martinfowler.com/bliki/TestCoverage.html
1. https://research.swtch.com/testing

It's what I used when doing a presentation on this a few months back. I considered posting the slides, but they are mostly just screenshots from the above resources.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 4 above)
Comment (jiridanek) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:329]: you've changed a tab-indented line into a 8-spaces indented line, try highlighting the marked portion in the diff with mouse, both original and the changed

usual go practice is to use tabs, ofc; if you run `go fmt`, it should fix it for you

**Comment by jiridanek** [comment]
> very nice, I will have to make myself more familiar with the findings. From your perspective, did I leave out anything by error in this PR when resolving diffs, or is all ok?
>
> I only took, like ClusterWideProxyIsEnabled, what someone else wrote wayyy back then ;-) Just Management Information Systems major, no code geek (yet) :-)

I don't know, would have to spend time with it. So far I had other things to mess around with. Among the repo maintainers, we're keeping this PR in mind, it's just nobody decided to take on the work to merge it in (and accept the responsibility for resolving any and all unforeseen problems that may cause ;)

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 4 above)
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go]: ah, of course, what a beginner error ...
make fmt / go fmt did the trick, thank you

**Comment by shalberd** [comment]
@jiridanek my aim is to test odh notebook controller image on one of our dev clusters in conjunctions with DataScienceCluster workbenches devFlags and ODH Operator 2.22.0.
The cluster will have
- no integrated Openshift registry
- custom imagestreams / notebooks
- a custom namespace tst-opendatahub
- a central proxy config at cluster-level

**Comment by shalberd** [comment]
@jiridanek @atheo89 @andrewballantyne I was just making my first steps with odh v2.22.0 operator and devFlags manifests download, could not figure out for the life of me why the creation of the folders in operator image at /opt/manifests did not work at all, meaning operator v.2.22.0 being unusable for anything devflag-related it seems like.
Found out why: https://github.com/opendatahub-io/opendatahub-operator/pull/1462
Don't know if Wen will backport the change to current v2.22.0 operator.

**Comment by shalberd** [comment]
> However, the upcoming ODH release with the v2.23 operator is scheduled for next week.
If that’s the case, will this PR still be relevant given the changes in the ODH operator?

@atheo89 yes, the fix for the operator has nothing do do with proxy-related env var injection into notebook main container. This PR here is still highly relevant for clusters that have proxy info in their default cluster proxy config.

The change in operator is related to: devFlags manifests download into operator container not working at all currently, so only related slightly, in the sense that I cannot test this custom odh notebook controller image here currently :-)

**Comment by shalberd** [comment]
@atheo89 @jiridanek 

I can confirm this is working withb the PR-326 version of odh-notebook-controller, tested via devFlags on OCP 4.15 with cluster proxy config.

Now, this is just from a functional perspective. It could well be, as our resolved comments showed before, that there are optimizations possible in the golang code.


proxy env injection working

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 5]
```diff
@@ -230,6 +233,29 @@ func InjectOAuthProxy(notebook *nbv1.Notebook, oauth OAuthConfig) error {
 	return nil
 }
 
+func (w *NotebookWebhook) ClusterWideProxyIsEnabled() bool {
+	proxyResourceList := &configv1.ProxyList{}
+	err := w.Client.List(context.TODO(), proxyResourceList)
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:238]: I'm looking to learn more about the Proxy CR, so far found https://docs.openshift.com/container-platform/4.17/rest_api/config_apis/proxy-config-openshift-io-v1.html. Just a note-to-self.

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 6]
```diff
@@ -369,6 +403,57 @@ func (w *NotebookWebhook) maybeRestartRunningNotebook(ctx context.Context, req a
 	return mutatedNotebook, &UpdatesPending{Reason: diff}, nil
 }
 
+func InjectProxyConfigEnvVars(notebook *nbv1.Notebook) error {
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:406]: There should be an envtest for this. When we decide to get the PR in, this is the first thing I'd like to look into.

We'd need to add a fake Proxy CRD, and have some test that starts Notebook with and without a proxy being configured. We don't much care for a scenario when Proxy CRD is missing (that's what we're testing now, actually, the `return false` when `err != nil` in `ClusterWideProxyIsEnabled` above) because OpenShift always has Proxy CRD.

**Comment by jiridanek** [comment]
This sounds promising. Last week or so, @harshad16 had a look and had some problems making everything fit together on his clusters. We did not discuss the details, but I understood it was already just the basic functionality (set proxy in openshift, have the env vars set in your notebook), that was not working right. We'll need to decide to get this into some future release and sort everything out. Eventually.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 6 above)
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:406]: feel free to push to my fork, if you want to. I am fine with that in my case.

**Comment by shalberd** [comment]
@jiridanek rebased to be in sync with main, in case you wanna try the built image, too.

**Review by harshad16 — APPROVED** [review]
/lgtm

Tested the PR, with following changes:
```yaml
apiVersion: config.openshift.io/v1
kind: Proxy
metadata:
  name: cluster
spec:
  httpProxy: http://httpbin.org
  httpsProxy: https://httpbin.org
  noProxy: localhost
```

Screenshot from 2025-02-25 00-50-19

It works as intended. thanks for the PR and patience.
I seriously appreciate your patience with me , in regards for testing the changes.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 6 above)
Comment (shalberd) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:406]: @jiridanek can you make a follow-up Story in Jira for this, please? Enhancements to code. 
I assume @harshad16  took this aspect into account when merging in as-is.
What I can say for certain: in daily operations, the controller is running just fine.

**Comment by shalberd** [comment]
will this make its way into v1.9.0-8 or -9 and a coming release of ODH, e.g. 2.25.0?

**Comment by shalberd** [comment]
will this make its way into v1.9.0-8 or -9 and a coming release of ODH, e.g. v2.25.0?
https://github.com/opendatahub-io/opendatahub-operator/releases

**Comment by jiridanek** [comment]
not -8, that went out without it, https://github.com/opendatahub-io/kubeflow/releases/tag/v1.9.0-8

should be in -9, so that's like ODH 2.26 I think
